dbms/src/Functions/FunctionsIsIPAddr.h

```diff
@@ -0,0 +1,288 @@
+            *tp = num;
+            if (!saw_digit)
+            {
+                if (++octets > 4)
+                    return 0;
+                saw_digit = 1;
+            }
+        }
+        else if (ch == '.' && saw_digit)
+        {
+            if (octets == 4)
+                return 0;
+            *++tp = 0;
+            saw_digit = 0;
+        }
+        else
+            return 0;
+    }
+    if (octets < 4)
+        return 0;
+
+    return 1;
+}
+
+/* Description：
+ * This function is used to determine whether the input string is an IPv6 address, 
+ * and the code comes from the inet_pton6 function of "arpa/inet.h".
+ * References: http://svn.apache.org/repos/asf/apr/apr/trunk/network_io/unix/inet_pton.c
+ */
+static inline UInt8 isIPv6(const char * src)
+{
+    if (nullptr == src)
+        return 0;
+    static const char xdigits_l[] = "0123456789abcdef",
+                      xdigits_u[] = "0123456789ABCDEF";
+    unsigned char tmp[16], *tp, *endp, *colonp;
+    const char *xdigits, *curtok;
+    int ch, saw_xdigit;
+    unsigned int val;
+
+    memset((tp = tmp), '\0', IN6ADDRSZ);
+    endp = tp + IN6ADDRSZ;
+    colonp = nullptr;
+    if (*src == ':')
+        if (*++src != ':')
+            return 0;
+    curtok = src;
+    saw_xdigit = 0;
+    val = 0;
+    while ((ch = *src++) != '\0')
+    {
+        const char * pch;
+
+        if ((pch = strchr((xdigits = xdigits_l), ch)) == nullptr)
+            pch = strchr((xdigits = xdigits_u), ch);
+        if (pch != nullptr)
+        {
+            val <<= 4;
+            val |= (pch - xdigits);
+            if (val > 0xffff)
+                return 0;
+            saw_xdigit = 1;
+            continue;
+        }
+        if (ch == ':')
+        {
+            curtok = src;
+            if (!saw_xdigit)
+            {
+                if (colonp)
+                    return 0;
+                colonp = tp;
+                continue;
+            }
+            if (tp + INT16SZ > endp)
+                return 0;
+            *tp++ = (unsigned char)(val >> 8) & 0xff;
+            *tp++ = (unsigned char)val & 0xff;
+            saw_xdigit = 0;
+            val = 0;
+            continue;
+        }
+        if (ch == '.' && ((tp + INADDRSZ) <= endp) && isIPv4(curtok) > 0)
+        {
+            tp += INADDRSZ;
+            saw_xdigit = 0;
+            break; /* '\0' was seen by isIPv4(). */
+        }
+        return 0;
+    }
+    if (saw_xdigit)
+    {
+        if (tp + INT16SZ > endp)
+            return 0;
+        *tp++ = (unsigned char)(val >> 8) & 0xff;
+        *tp++ = (unsigned char)val & 0xff;
+    }
+    if (colonp != nullptr)
+    {
+        const size_t n = tp - colonp;
+        size_t i;
+
+        for (i = 1; i <= n; i++)
+        {
+            endp[-i] = colonp[n - i];
+            colonp[n - i] = 0;
+        }
+        tp = endp;
+    }
+    if (tp != endp)
+        return 0;
+    return 1;
+}
+
+class FunctionIsIPv4 : public IFunction
```

Comment:
We can use policy trait to create a class like：
```
template <bool isIPv4>
class FunctionIsIPv4OrIPv6 : public IFunction
```
Then you can use constexpr to check whether to use `isIPv4` or `isIPv6`

Reply:
🤔  why? They are different functions.

Reply:
> 🤔 why? They are different functions.

But the basic implementation logic is same, we can reuse the code.

Reply:
> > 🤔 why? They are different functions.
> 
> But the basic implementation logic is same, we can reuse the code without any performance impact.

Reply:
but not clear

Reply:
> but not clear

Found many examples in dbms/src/Functions/FunctionsURL.cpp

Reply:
Thanks a lot, I referenced other functions. The following template is used
`template <typename Name, UInt8(Function)(const char *)>`